### PR TITLE
Concat

### PR DESCRIPTION
**Pull request by pbrunet**

### CODE REVIEW

File: pythran/pythonic/numpy/concatenate.hpp [hunk 1]
```diff
@@ -27,22 +28,49 @@ namespace pythonic {
...
+                typename types::numpy_expr_to_ndarray<A>::type t = arr; // We force evaluation of the ndarray
+                std::copy(t.fbegin(), t.fend(), iter);
             }
-        template<class T, size_t N, class A, size_t M>
-            void _concatenate_helper(types::array<T, N>& a, A const& t, utils::int_<M>) {
-                a[N - M] = std::get<N-M>(t);
-                _concatenate_helper(a, t, utils::int_<M-1>());
+
+        template<class A, size_t M, class ...Types>
+            void _concatenate_helper(std::tuple<Types...> const& a, A const& arr, typename assignable<typename __combined<Types...>::type>::type::dtype* iter, utils::int_<M>);
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/concatenate.hpp]: this forward declaration is no loner useful, I guess.

File: pythran/pythonic/numpy/concatenate.hpp [hunk 2]
```diff
@@ -27,22 +28,49 @@ namespace pythonic {
...
+
+        template<class A, size_t M, class ...Types>
+            void _concatenate_helper(std::tuple<Types...> const& a, A const& arr, typename assignable<typename __combined<Types...>::type>::type::dtype* iter, utils::int_<M>) {
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/concatenate.hpp]: The second argument is useless, you could recover the Mth element in the body.

File: pythran/pythonic/numpy/concatenate.hpp [hunk 3]
```diff
@@ -27,22 +28,49 @@ namespace pythonic {
...
+            {
+                // FIXME This is a hack to force list.fake_shape evaluation to types::array
+                return fake_shape[0];
             }
+
+        template<class... Types, int... I>
+        std::tuple<size_t, long> concatenate_size_helper(std::tuple<Types...> const& args, utils::seq<I...>)
+        {
+            return std::tuple<size_t, long>{__builtin__::sum(std::make_tuple(std::get<I-1>(args).flat_size()...)), __builtin__::sum(std::make_tuple(first_dim<std::tuple_element<0, std::tuple<Types...>>::type::value>(std::get<I-1>(args).shape)...))};
+        }
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/concatenate.hpp]: all these helpers should go in a details namespace.

**Comment by pbrunet** [comment]
This one is OK

**Pull request merged**